File: node/src/activation.ts
import { HttpClient } from './http-client';
import type { ShipzyConfig } from './http-client';
import type { ApiResult } from './index';

// ============ Types ============

export type Capability = 'signature' | 'timestamp' | 'identity';

export interface Provider {
    slug: string;
    name: string;
    capabilities: Capability[];
    status: string;
}

export interface ProviderActivation {
    id: string;
    provider_slug: string;
    merchant_id: string;
    status: string;
    created_at: string;
    updated_at: string;
}

export interface ActivateRequest {
    provider_slug: string;
    [key: string]: any;
}

// ============ Client ============

export class ActivationClient extends HttpClient {
    async listProviders(filter?: { capability?: Capability }): Promise<ApiResult<{ data: Provider[] }>> {
        const q = filter?.capability ? `?capability=${encodeURIComponent(filter.capability)}` : '';
        return this.request(`/api/v1/marketplace/providers${q}`);
    }

    async getProvider(slug: string): Promise<ApiResult<Provider>> {
        return this.request(`/api/v1/marketplace/providers/${encodeURIComponent(slug)}`);
    }

    async list(): Promise<ApiResult<{ data: ProviderActivation[] }>> {
        return this.request('/api/v1/marketplace/activations');
    }

    async get(id: string): Promise<ApiResult<ProviderActivation>> {
        return this.request(`/api/v1/marketplace/activations/${encodeURIComponent(id)}`);
    }

    async activate(req: ActivateRequest): Promise<ApiResult<ProviderActivation>> {
        return this.request('/api/v1/marketplace/activations', 'POST', req);
    }

    async pause(id: string): Promise<ApiResult<{ status: string }>> {
        return this.request(`/api/v1/marketplace/activations/${encodeURIComponent(id)}/pause`, 'POST', {});
    }

    async resume(id: string): Promise<ApiResult<{ status: string }>> {
        return this.request(`/api/v1/marketplace/activations/${encodeURIComponent(id)}/resume`, 'POST', {});
    }

    async revoke(id: string, reason?: string): Promise<ApiResult<{ status: string }>> {
        return this.request(`/api/v1/marketplace/activations/${encodeURIComponent(id)}/revoke`, 'POST', { reason: reason || '' });
    }
}

File: python/zymeup_sdk/activation.py
"""Zymeup Python SDK - Activation Client"""
from typing import Optional, Dict, Any, List


class ActivationClient:
    """Provider activation client"""

    def __init__(self, client):
        self._client = client

    def list_providers(self, capability: Optional[str] = None) -> Dict[str, Any]:
        """List available providers"""
        params: Dict[str, Any] = {}
        if capability:
            params["capability"] = capability
        return self._client.request("GET", "/api/v1/marketplace/providers", params=params)

    def get_provider(self, slug: str) -> Dict[str, Any]:
        """Get provider detail"""
        return self._client.request("GET", f"/api/v1/marketplace/providers/{slug}")

    def list(self) -> Dict[str, Any]:
        """List merchant's activations"""
        return self._client.request("GET", "/api/v1/marketplace/activations")

    def get(self, activation_id: str) -> Dict[str, Any]:
        """Get activation detail"""
        return self._client.request("GET", f"/api/v1/marketplace/activations/{activation_id}")

    def activate(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Activate a provider"""
        return self._client.request("POST", "/api/v1/marketplace/activations", json=data)

    def pause(self, activation_id: str) -> Dict[str, Any]:
        """Pause an activation"""
        return self._client.request("POST", f"/api/v1/marketplace/activations/{activation_id}/pause")

    def resume(self, activation_id: str) -> Dict[str, Any]:
        """Resume an activation"""
        return self._client.request("POST", f"/api/v1/marketplace/activations/{activation_id}/resume")

    def revoke(self, activation_id: str, reason: Optional[str] = None) -> Dict[str, Any]:
        """Revoke an activation"""
        return self._client.request("POST", f"/api/v1/marketplace/activations/{activation_id}/revoke", json={"reason": reason or ""})
